revision.c

```diff
@@ -624,11 +625,70 @@ static void file_change(struct diff_options *options,
 	options->flags.has_changes = 1;
 }
 
+static struct trace_key trace_bloom_filter = TRACE_KEY_INIT(BLOOM_FILTER);
+static int trace_bloom_filter_atexit_registered;
+static unsigned int bloom_filter_count_maybe;
+static unsigned int bloom_filter_count_definitely_not;
+static unsigned int bloom_filter_count_false_positive;
+static unsigned int bloom_filter_count_not_parsed;
+
+static void print_bloom_filter_stats_atexit(void)
+{
+    unsigned int total = bloom_filter_count_maybe +
+	bloom_filter_count_definitely_not;
+
+	trace_printf_key(&trace_bloom_filter,
+			 "bloom filter total queries: \t%d \tdefinitely not: \t%d \tmaybe:  \t%d \tfalse positives: \t%d \tnot_parsed: \t%d \tfp ratio: \t%f\n",
+			 total,
+			 bloom_filter_count_definitely_not,
+			 bloom_filter_count_maybe,
+			 bloom_filter_count_false_positive,
+			 bloom_filter_count_not_parsed,
+			 (1.0 * bloom_filter_count_false_positive) / total);
```

Comment:
Replace with `trace2_data_intmax()` as in [midx.c](https://github.com/git/git/blob/ad05a3d8e5a6a06443836b5e40434262d992889a/midx.c#L170-L171) or [read-cache.c](https://github.com/git/git/blob/ad05a3d8e5a6a06443836b5e40434262d992889a/read-cache.c#L2257-L2260). It looks like the stats I tracked for the topo-order code didn't make it into the product as I had thought.

Reply:
Maybe only report the stats you need to derive the rest:

```
bloom_filter_count_definitely_not
bloom_filter_count_maybe
bloom_filter_count_false_positive
bloom_filter_count_not_parsed
```

Reply:
@derrickstolee  trace2_printf() only go to the PERF log, not the EVENT log.  trace2_data_*() messages to go both.